recipes/dftd4/dftd4-openmp-gcc7.patch

```diff
@@ -0,0 +1,94 @@
+diff --git a/source/dftd4.f90 b/source/dftd4.f90
+index e6a1dae..726c1e7 100644
+--- a/source/dftd4.f90
++++ b/source/dftd4.f90
+@@ -163,7 +163,7 @@ subroutine initialize_d4_model(self,atoms,mode,g_a,g_c)
+    enddo
+ 
+    ! integrate C6 coefficients
+-   !$omp parallel default(none) private(i,j,ii,jj,alpha,c6) shared(self)
++   !$omp parallel default(shared) private(i,j,ii,jj,alpha,c6) shared(self)
+    !$omp do schedule(runtime)
+    do i = 1, max_elem
+       do j = 1, i
+@@ -232,7 +232,7 @@ subroutine initialize_d3_model(self,atoms)
+    enddo
+ 
+    ! integrate C6 coefficients
+-   !$omp parallel default(none) private(i,j,ii,jj,alpha,c6) shared(self)
++   !$omp parallel default(shared) private(i,j,ii,jj,alpha,c6) shared(self)
+    !$omp do schedule(runtime)
+    do i = 1, max_elem
+       do j = 1, i
+@@ -1195,7 +1195,7 @@ subroutine edisp_3d(mol,dispm,ndim,q,r_thr,mbd_thr,par, &
+    enddo
+ 
+ !  print'(" * Entering first OMP section")'
+-!$OMP parallel default(none) &
++!$OMP parallel default(shared) &
+ !$omp private(i,ii,iii,ia,iz,k)  &
+ !$omp shared (dispm,mol,itbl,q) &
+ !$omp shared (gw,zetavec,zerovec,r_thr)
+@@ -1214,7 +1214,7 @@ subroutine edisp_3d(mol,dispm,ndim,q,r_thr,mbd_thr,par, &
+ !$omp end do
+ !$omp end parallel
+ 
+-!$OMP parallel default(none) &
++!$OMP parallel default(shared) &
+ !$omp private(i,j,ia,ja,ij,k,l,c6ii,c6ij,disp,  &
+ !$omp         rij,r2,r,r4r2ij,r0,oor6,oor8,oor10, &
+ !$omp         t,tx,ty,tz)  &
+@@ -1329,7 +1329,7 @@ subroutine abcappr_3d(mol,dispm,ndim,par,zvec,refc6,itbl,rep,r_thr,eabc)
+ 
+    eabc = 0.0_wp
+ 
+-!$OMP parallel default(none) &
++!$OMP parallel default(shared) &
```

Comment:
This is a potentially dangerous thing to do in a patch BTW. If you are an expert in this code base and openmp, then ignore me. Otherwise, i'd hold off on defaulting variables to shared for far of a race condition.

Reply:
This seems to be a problem with constants in OMP and GCC7 when `default(none)` is used. The only way to fix this is using a newer GCC version or by switching from `default(none)` to `default(shared)`, I chose the latter here.

Reply:
Not the optimal solution but better than no OMP.

Reply:
I don't agree. As an experience openmp user, this is a bad idea. Let's change the patch to explicitly add the constants not listed to the shared directive.

Reply:
This is actually the point, you can't add the constants explicitly. Compiling with `default(none)` and GCC7 gives:
```
../source/dftd4.f90:2232:0:

             cij  = par%a1*sqrt(3._wp*r4r2(at(iat))*r4r2(at(jat)))+par%a2
 
Error: ‘sqrt_z_r4_over_r2’ not specified in enclosing ‘parallel’
```
adding it explicitly returns an error as well:
```
../source/dftd4.f90:2210:52:

 !$omp shared(nat,xyz,c6ab,rep,par,at,dlat,dc6ab,thr,r4r2) &
                                                    1
Error: Object ‘sqrt_z_r4_over_r2’ is not a variable at (1)
```

Reply:
The constant is renamed with `r4r2 => sqrt_z_r4_over_r2` in the use statement, but the same happens when not renaming it (just tried it again to be sure).

Reply:
Can we try adding them as firstprivate? 

https://gcc.gnu.org/gcc-9/porting_to.html#ompdatasharing

Then i'll give up.

Reply:
Doesn't work either. It only happens with GCC7 and lower, GCC8 and Intel have it fixed.
```
../source/dftd4.f90:2211:19:

 !$omp firstprivate(r4r2) &
                   1
Error: Object ‘sqrt_z_r4_over_r2’ is not a variable at (1)
```

Reply:
frack ok